controllers/nifi_controller.go

```diff
@@ -0,0 +1,1092 @@
+/*
+Copyright 2022.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/resource"
+	resourcev1 "k8s.io/apimachinery/pkg/api/resource"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	bigdatav1alpha1 "github.com/RHEcosystemAppEng/nifi-operator/api/v1alpha1"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+var (
+	serviceName = "cluster"
+	namespace   = "nifi-operator"
+)
+
+const (
+	nifiPrefix = "nifi-"
+	// nifiImageRepo sets the repo URL for the Nifi image
+	nifiImageRepo = "docker.io/apache/nifi"
+	// nifiVersion sets the version for the Nifi Image
+	nifiVersion = ":1.16.3"
+	// nifiUser internal user ID for nifi
+	nifiUser = int64(1000)
+	// nifiPropertiesAccessMode establish the internal unix permissions for 'nifi.properties' file
+	nifiPropertiesAccessMode = int32(420)
+	// nifiDefaultUser sets Single User Access Username
+	nifiDefaultUser = "administrator"
+	// nifiDefaultUser sets Single User Access Password
+	nifiDefaultPassword = "administrator"
```

Comment:
The default password should be secure random generated and stored in a secret then loaded into the container.

Reply:
great feedback. I replaced the nifiDefgaultPassword by a random generated one, and save it in a Secret. Secret content is not reconciled to avoid password re-generation every time